frame/merkle-mountain-range/src/lib.rs

```diff
@@ -334,6 +334,31 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		mmr.generate_batch_proof(leaf_indices)
 	}
 
+	/// Generate a MMR proof for the given `leaf_indices` at the `best_known_block`.
+	///
+	/// Note this method can only be used from an off-chain context
+	/// (Offchain Worker or Runtime API call), since it requires
+	/// all the leaves to be present.
+	/// It may return an error or panic if used incorrectly.
+	pub fn generate_historical_batch_proof(
+		leaf_indices: Vec<LeafIndex>,
+		best_known_block: <T as frame_system::Config>::BlockNumber,
+	) -> Result<
+		(Vec<LeafOf<T, I>>, primitives::BatchProof<<T as Config<I>>::Hash>),
+		primitives::Error,
+	>
+	where
+		<T as frame_system::Config>::BlockNumber: Into<LeafIndex>,
+	{
+		let block_age = <frame_system::Pallet<T>>::block_number()
```

Comment:
I'd use `checked_sub` here and return err on `None`. the caller may provide valid future `best_known_block` that is missing from our local chain and we'll generate proof that he won't be able to verify using mmr hash at `best_known_block`.

Reply:
I changed the method's signature to work with leaf indices and this doesn't apply anymore. But it's a good catch ! I'll keep this in mind if we change the API to work with block numbers instead of leaf indices.